File: C4/4.3.py
# 4.3 Given a binary tree, design an algorithm which creates a linked list of
# all nodes at each depth

class BinaryTreeNode():
	def __init__(self, val):
		self.Value = val
		self.Left = None
		self.Right = None

class LinkedListNode():
	def __init__(self, val):
		self.Value = val
		self.Next = None




rooty = BinaryTreeNode(1)
rooty.Left = BinaryTreeNode(2)
rooty.Right = BinaryTreeNode(3)
rooty.Left.Left = BinaryTreeNode(4)
rooty.Left.Right = BinaryTreeNode(5)
rooty.Right.Left = BinaryTreeNode(6)
rooty.Right.Right = BinaryTreeNode(7)

def LinkedListAtDepths(rootNode):
	# create queue that initially has just the rootnode
	# we use BF iteration to append items onto a list
	# create linkedList from those items
	# also have a temporary array that has the children from that BF iteration
	# reset the queue to the array of children nodes
	queue = [rootNode]
	returnArrOfLinkedListNodes = []
	while len(queue) > 0:
		nextQueue = []
		llHead = None
		llNode = None
		while len(queue) > 0:
			item = queue.pop(0)
			if item.Left != None:
				nextQueue.append(item.Left)
			if item.Right != None:
				nextQueue.append(item.Right)
			if llHead == None:
				llNode = LinkedListNode(item.Value)
				llHead = llNode
			else:
				llNode.Next = LinkedListNode(item.Value)
				llNode = llNode.Next

		returnArrOfLinkedListNodes.append(llHead)
		if len(nextQueue) > 0:
			queue = nextQueue

	return returnArrOfLinkedListNodes


testArr = LinkedListAtDepths(rooty)
for i in range(len(testArr)):
	node = testArr[i]
	while node != None:
		print(node.Value)
		node = node.Next

File: C4/4.3.go
/*
Objec­tive: Given a Binary tree cre­ate Linked Lists of all the nodes at each depth , say if the tree has height k then cre­ate k linked lists.

List of Depths: Given a binary tree, design an algorithm which creates a linked list of all the nodes at each depth
(e.g., if you have a tree with depth x, you'll have x linked lists).

I: Binary Tree (maybe not BST)
O: Linked Lists (n) where n is also the depth
C: None
E:
Unbalanced Tree?

Just one branch?
- a bunch of one node linkedLists?

Root is just one node?
*/

package main

import (
  "fmt"
)

type LinkedListNode struct {
  Val int
  Next *LinkedListNode
}

type BinaryTreeNode struct {
  Val int
  Left *BinaryTreeNode
  Right *BinaryTreeNode
}

// func (bst *BinaryTreeNode) LinkedListDepth() []*LinkedListNode {
//
// }
func inOrderTraversal(btn *BinaryTreeNode) {
  // does not need to be a pointer method since we're just printing nodes
  if btn != nil {
    inOrderTraversal(btn.Left)
    fmt.Println(btn.Val)
    inOrderTraversal(btn.Right)
  }
}

func main() {
  bst := &BinaryTreeNode{Val: 1}
  // create a pointer to BinaryTreeNode that has value of 1
  // new is equivalent to this
  // var bst *BinaryTreeNode is different in that it offers a nil value
  // pointer and new saves space in memory with nil values (different!)
  bst.Left = &BinaryTreeNode{Val: 2}
  bst.Right = &BinaryTreeNode{Val: 3}
  bst.Left.Left = &BinaryTreeNode{Val: 4}
  bst.Left.Right = &BinaryTreeNode{Val: 5}
  bst.Right.Left = &BinaryTreeNode{Val: 6}
  bst.Right.Right = &BinaryTreeNode{Val: 7}
  inOrderTraversal(bst)
}
